### PR TITLE
input_shaper: Initial support of input shaping

### PR DESCRIPTION
**Pull request by dmbutyugin**
Input shaping can help to reduce printer vibrations due to resonances
and eliminate or reduce ghosting in prints.

Signed-off-by: Dmitry Butyugin <dmbutyugin@google.com>

### CODE REVIEW

File: klippy/extras/input_shaper.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+                or shaper_type_y != self.shaper_type_y):
+            self.toolhead.flush_step_generation()
+        ffi_main, ffi_lib = chelper.get_ffi()
+        new_delay = max(
+                ffi_lib.input_shaper_get_step_generation_window(
+                    shaper_type_x, shaper_freq_x, damping_ratio_x),
+                ffi_lib.input_shaper_get_step_generation_window(
+                    shaper_type_y, shaper_freq_y, damping_ratio_y))
+        self.toolhead.note_step_generation_scan_time(new_delay,
+                                                     old_delay=self.old_delay)
```
Comment (KevinOConnor) [inline klippy/extras/input_shaper.py:79]: `self.old_delay` needs to be updated afterwards.

File: klippy/kinematics/cartesian.py [hunk 2]
```diff
@@ -126,7 +126,7 @@ def _activate_carriage(self, carriage):
         toolhead.flush_step_generation()
         dc_rail = self.dual_carriage_rails[carriage]
         dc_axis = self.dual_carriage_axis
-        self.rails[dc_axis].set_trapq(None)
+        self.dual_carriage_rails[not carriage].set_trapq(None)
```
Comment (KevinOConnor) [inline klippy/kinematics/cartesian.py:129]: I don't understand this change.  Can you elaborate on what it does?

Also, I suspect it should be in a separate commit (that way it if there is a problem it does not complicate a "git revert").

**Comment by KevinOConnor** [comment]
As a high-level comment, I'd like to get your thoughts on configuring the "frequency in hz" vs "period in seconds".  I'm fine with either, but think it would be good to discuss the pros and cons.

For frequency, I'm guessing the main pro is that the literature uses frequency.  It's also a little easier to discuss larger numbers (eg, "55hz" vs ".018s").

For seconds, it may be easier to correlate an increase in period with an increase in distance between "ridges on a print".  (Said another way, it's a little awkward to document that a decrease in distance between ridges calls for an increase in hz.)  I suspect it's also easier to describe the impact of "rounding of the corners" in terms of seconds than in hz.  (It could be awkward to document that a decrease in configured hz results in an increase in rounding.)

How does "range of suppression" of the various shaper types work in hz vs seconds?  For example, is it easier to describe (in the general case) the applicable range of zv vs ei in one particular set of units?

-Kevin

**Comment by KevinOConnor** [comment]
FYI, I have a few minor high-level comments on the code.  I think we can address these comments (if we agree) after merging.

1. Is there significant value in the zvd, zvdd, zvddd, and ei3hump shapers?  I got the impression that ei is going to be superior to the zvd variants in practical cases.  I got the impression that ei3hump is going to cause too much rounding to be useful in real world cases.  I understand the value in these for testing, but for the main code it might be better to focus on those shapers that we have good testing feedback with.  (Said another way, including support for lesser used (and less valuable) shapers may add confusion to users as they wade through documentation, and it may increase the burden on developers if they have to wade through "dead" code.)

2. FWIW, I found the `calc_position()` / `move_get_axis_coord()` code to be hard to understand.  I think it might be worthwhile to implement `calc_position()` with something like `for (i=0; i<n; i++) res += pulses[i].a * get_axis_position_across_moves(m, axis, m->move_t + pulses[i].t)`.  I can send a cumulative patch for discussion.

-Kevin

File: docs/G-Codes.md [hunk 3]
```diff
@@ -596,3 +596,18 @@ been enabled:
     delay duration for the identified [delayed_gcode] and starts the timer
     for gcode execution.  A value of 0 will cancel a pending delayed gcode
     from executing.
+
+## Resonance compensation
+
+The following command is enabled if an [input_shaping] config section has
```
[the remark this answers is not in the document]
Comment (dmbutyugin) [reply docs/G-Codes.md]: Done

File: klippy/extras/input_shaper.py (hunk 1 above)
Comment (dmbutyugin) [reply klippy/extras/input_shaper.py:79]: Good catch!

File: klippy/kinematics/cartesian.py (hunk 2 above)
Comment (dmbutyugin) [reply klippy/kinematics/cartesian.py:129]: Yes. This is actually a work-around of an issue with dual carriage support initially reported for axes smoothing, but also existing for input shaping. The problematic part is [this one](https://github.com/KevinOConnor/klipper/pull/3032/files#diff-7e6a0286c1f3bbf7493501136a9f7f82R58
):
+            s.set_trapq(self.toolhead.get_trapq())

This code resets trapq on all steppers on connection, including deactivated dual carriage stepper. Now the previous version of the code has self.rails[dc_axis] == dc_rail on initial execution of _activate_carriage. So essentially the previous version of the code did nothing useful here on the first call: it would set dc_rail trapq to None and then immediately to toolhead.get_trapq(). So I took advantage of it and made sure that it always resets trapq of a carriage being deactivated, which also fixes the issue in input_shaper.

Another alternative could be to add an API call to a stepper class to return its active trapq - so then input_shaper code can check whether it's None or not. Would that be a better alternative? And I wonder if it will have other side-effects.

I agree that how it is now, it may be a good idea to commit it separately. If sticking with this approach, would it be better to commit it before this pull request?

**Comment by dmbutyugin** [comment]
> FYI, I have a few minor high-level comments on the code. I think we can address these comments (if we agree) after merging.
> 
>     1. Is there significant value in the zvd, zvdd, zvddd, and ei3hump shapers?  I got the impression that ei is going to be superior to the zvd variants in practical cases.  I got the impression that ei3hump is going to cause too much rounding to be useful in real world cases.  I understand the value in these for testing, but for the main code it might be better to focus on those shapers that we have good testing feedback with.  (Said another way, including support for lesser used (and less valuable) shapers may add confusion to users as they wade through documentation, and it may increase the burden on developers if they have to wade through "dead" code.)

For ZVDD, ZVDDD I, in principle, agree. One may even argue that there is not much value in ZVD and EI should always be preferred instead. But I think there is a lot of value in 2hump and 3hump EI shapers. However, most of the time where this value exists, they should be configured differently than shaper_freq == resonance_freq (e.g. on your printer there seems to be resonances as low as 10-15 Hz, but one shouldn't use that value as a shaper_freq for 2hump and 3hump EI shapers). I need to write my thoughts in a more lengthy reply regarding the shapers API (frequency vs. time) and I am going to touch this point there too.

If you prefer, I can delete ZVDD and ZVDDD shapers. I can even delete ZVD if you insist, but that one is not too much maintenance burden, I'd say.

>     2. FWIW, I found the `calc_position()` / `move_get_axis_coord()` code to be hard to understand.  I think it might be worthwhile to implement `calc_position()` with something like `for (i=0; i<n; i++) res += pulses[i].a * get_axis_position_across_moves(m, axis, m->move_t + pulses[i].t)`.  I can send a cumulative patch for discussion.

I see your point. Well, `move_get_axis_coord()` is probably straightforward. But the reason for a more complex implementation of `calc_position()` is to calculate 'shaped' position in a single pass over moves. Simplify3D is known to generate lots of small moves, and one some prints the implementation with `get_axis_position_across_moves`, which has quadratic complexity, will get slower, potentially reaching the performance capabilities of a lower-end devices. I cannot tell how much of a problem that would really be - I can only say that I tried measuring vibration response of a printer with input shapers enabled, and I do run into some performance issues even with the current implementation on shapers with many pulses (e.g. 2hump and 3 hump shapers) - but this may be a not very realistic scenario in practice because vibration testing uses very fine segmentation now with moves having 0.0005 seconds duration (so an input shaper can easily span 50-80 moves).

> FWIW, could this "0 frequency" patch be replaced with a "dummy shaper" with "a=1.0, t=0.0, n=1"?

It could be, yes. However, there's already a code for short-circuiting the calculations based on active_flags, so I wonder if that would simplify things, unless we decide to fully use that dummy shaper consistently for that case as well.

Please let me know what you think about these points and my answers.

File: klippy/kinematics/cartesian.py (hunk 2 above)
Comment (KevinOConnor) [reply klippy/kinematics/cartesian.py:129]: Ah, okay.  I think we could enhance `stepper.set_stepper_kinematics()` to automatically reregister the existing trapq (if valid).  I think input_shaper.py then wouldn't need to call `set_trapq()` at all.

-Kevin

**Comment by KevinOConnor** [comment]
> If you prefer, I can delete ZVDD and ZVDDD shapers. I can even delete ZVD if you insist, but that one is not too much maintenance burden, I'd say.

These are just points for discussion.  If you feel they have value then that's fine with me.  I don't think there is significant code burden to them.

> Edit: The potential implementation with get_axis_position_across_moves does not have a quadratic complexity after some thought.

FWIW, I'd have guessed that `get_axis_position_across_moves` would be basically the same speed, but I haven't looked closely.

> However, there's already a code for short-circuiting the calculations based on active_flags

Good point.  I agree no change is needed.

-Kevin

**Comment by dmbutyugin** [comment]
@KevinOConnor, I'll delete ZVDD and ZVDDD input shapers for now. TBH, I do not see where I could suggest to use them. And if somebody really requests them, they can be easily added back.

> FWIW, I'd have guessed that get_axis_position_across_moves would be basically the same speed, but I haven't looked closely.

Yes, I think it makes sense to simplify the implementation. I'll make a change.

File: klippy/kinematics/cartesian.py (hunk 2 above)
Comment (dmbutyugin) [reply klippy/kinematics/cartesian.py:129]: I do not know if that would interfere with [force_move](https://github.com/KevinOConnor/klipper/blob/master/klippy/extras/force_move.py#L82). I could probably use a snippet there:
```
prev_trapq = stepper.set_trapq(self.trapq)
prev_sk = stepper.set_stepper_kinematics(self.stepper_kinematics)
...
stepper.set_trapq(prev_trapq)
stepper.set_stepper_kinematics(prev_sk)
```

Still, do you prefer to do that refactoring in a separate pull request? Before this one?

Comment (KevinOConnor) [reply klippy/kinematics/cartesian.py:129]: Yes - that's how I would fix force_move.  Probably simplest to commit the change to set_stepper_kinematics first.  I can make that change if you like - I believe it to be:
```diff
--- a/klippy/extras/force_move.py
+++ b/klippy/extras/force_move.py
@@ -39,7 +39,6 @@ class ForceMove:
         self.trapq_free_moves = ffi_lib.trapq_free_moves
         self.stepper_kinematics = ffi_main.gc(
             ffi_lib.cartesian_stepper_alloc('x'), ffi_lib.free)
-        ffi_lib.itersolve_set_trapq(self.stepper_kinematics, self.trapq)
         # Register commands
         gcode = self.printer.lookup_object('gcode')
         gcode.register_command('STEPPER_BUZZ', self.cmd_STEPPER_BUZZ,
@@ -80,6 +79,7 @@ class ForceMove:
         toolhead = self.printer.lookup_object('toolhead')
         toolhead.flush_step_generation()
         prev_sk = stepper.set_stepper_kinematics(self.stepper_kinematics)
+        prev_trapq = stepper.set_trapq(self.trapq)
         stepper.set_position((0., 0., 0.))
         axis_r, accel_t, cruise_t, cruise_v = calc_move_time(dist, speed, accel)
         print_time = toolhead.get_last_move_time()
@@ -88,6 +88,7 @@ class ForceMove:
         print_time = print_time + accel_t + cruise_t + accel_t
         stepper.generate_steps(print_time)
         self.trapq_free_moves(self.trapq, print_time + 99999.9)
+        stepper.set_trapq(prev_trapq)
         stepper.set_stepper_kinematics(prev_sk)
         toolhead.note_kinematic_activity(print_time)
         toolhead.dwell(accel_t + cruise_t + accel_t)
diff --git a/klippy/stepper.py b/klippy/stepper.py
index 4249dccb..ad3a6006 100644
--- a/klippy/stepper.py
+++ b/klippy/stepper.py
@@ -124,8 +124,9 @@ class MCU_stepper:
         old_sk = self._stepper_kinematics
         self._stepper_kinematics = sk
         if sk is not None:
-            self._ffi_lib.itersolve_set_stepcompress(
-                sk, self._stepqueue, self._step_dist)
+            self._ffi_lib.itersolve_set_stepcompress(sk, self._stepqueue,
+                                                     self._step_dist)
+            self.set_trapq(self._trapq)
         return old_sk
     def note_homing_end(self, did_trigger=False):
         ret = self._ffi_lib.stepcompress_reset(self._stepqueue, 0)
```

-Kevin

**Comment by dmbutyugin** [comment]
> As a high-level comment, I'd like to get your thoughts on configuring the "frequency in hz" vs "period in seconds". I'm fine with either, but think it would be good to discuss the pros and cons.

My (not very strong) preference is to use frequency in Hz instead of periods in seconds.

> For frequency, I'm guessing the main pro is that the literature uses frequency. It's also a little easier to discuss larger numbers (eg, "55hz" vs ".018s").

I agree with this. Also, if we ever use FFT for resonance measurements, it also naturally reports frequency instead of periods.

> For seconds, it may be easier to correlate an increase in period with an increase in distance between "ridges on a print". (Said another way, it's a little awkward to document that a decrease in distance between ridges calls for an increase in hz.)

There is a simple formula in the doc which gives direct instructions what to multiple and divide by what to get a result. But I think also the following rule of thumb may help explain this: higher frequency means faster oscillations, and faster the oscillations - the more ridges 'fit' into the test.
 

> I suspect it's also easier to describe the impact of "rounding of the corners" in terms of seconds than in hz. (It could be awkward to document that a decrease in configured hz results in an increase in rounding.)

It is very possible that it's just me, but I don't personally have a good 'hunch' as to which numbers for shaper are 'still good' and which ones are 'too large'. I.e. by just looking at the number, is 0.02 sec too large for EI in terms of smoothing the corners? What about 2hump EI? 3hump EI? I wouldn't be able to tell from the top of my head. I.e. 0.02 sec for EI and 3-hump EI shapers are very not the same. So it may actually mislead the users into thinking that the same period for different shapers == same smoothing, which is wrong.

Then another confusing part (for me at least) is that smoothing grows quadratically with increase of a period (to be more precise, there is a linear component proportional to square_corner_velocity / shaper_freq and quadratic proportional to accel / shaper_freq^2). So, incidentally, 0.05 sec and 0.04 sec period will be quite different, even though they look relatively close (~56% smoothing increase instead of only 25%). 

All I am saying is that for me shaper period in seconds does not add much intuition. But then again, it may be just me. It would also be good to hear other peoples' opinions. @BlackStump what did you find easier to specify and understand, Hz or sec when configuring smooth_axis earlier and input_shaper now?


> How does "range of suppression" of the various shaper types work in hz vs seconds? For example, is it easier to describe (in the general case) the applicable range of zv vs ei in one particular set of units?

We can say, for example, that EI shaper reduces vibrations by 95% in the frequency range +/- 20% from the configured frequency (I want to make a table giving that info for each supported shaper). So if a user wants to get which frequencies their EI shaper configured for 38 Hz suppresses, they need to multiply that number by 0.8 and 1.2 (so it's [30.4, 45.6] Hz range). Since the shapers are symmetric in frequency, it's the same rule for shaper period: EI shaper at 0.0263 sec period reduces vibrations in range of [0.02104, 0.03156]. Is it sufficiently illustrative?

BTW, I specifically didn't name the parameters 'resonance_freq_x' and alike, because it should be a resonance frequency only for shapers like ZV and MZV (well, and usually EI shaper). 2-hump EI and 3-hump EI shapers should usually be configured differently. I also want to introduce another parameter for them `min_resonance_freq_x/y` - so that a user could put that value into a config and then be able to switch between N-hump EI shapers without having to do this math exercises of converting shaper_freq parameter between them. But I won't do it in this pull request, otherwise there's often 'one more thing to improve' and it can get some time to get it right.

File: klippy/kinematics/cartesian.py (hunk 2 above)
Comment (dmbutyugin) [reply klippy/kinematics/cartesian.py:129]: Also perhaps extruder.py will need to be updated? But sure, I don't mind either way.

**Comment by BlackStump** [comment]
Hz would be my choice

File: klippy/kinematics/cartesian.py (hunk 2 above)
Comment (KevinOConnor) [reply klippy/kinematics/cartesian.py:129]: Okay - I committed the change (c6b65204).  I don't think extruder.py needs to change.

-Kevin

**Comment by KevinOConnor** [comment]
> It is very possible that it's just me, but I don't personally have a good 'hunch' as to which numbers for shaper are 'still good' and which ones are 'too large'. I.e. by just looking at the number, is 0.02 sec too large for EI in terms of smoothing the corners? What about 2hump EI? 3hump EI? I wouldn't be able to tell from the top of my head. I.e. 0.02 sec for EI and 3-hump EI shapers are very not the same. So it may actually mislead the users into thinking that the same period for different shapers == same smoothing, which is wrong.

I guess I was thinking that it would be easier to document the total "shaping time".  So, a ".020s zv shaper" is done over 10ms, ".020s mzv" is 15ms, ".020s ei" is 20ms, ".020s 2hump_ei" is 30ms.  Admittedly, it's unclear if ".020s ei" produces more or less rounding than ".040s zv", and it's unclear if there is a consistent "shaping time" that we could broadly say is too high.

-Kevin

**Comment by dmbutyugin** [comment]
@KevinOConnor I've made the suggested changes. Please take a look at the updated code.

File: docs/G-Codes.md [hunk 4]
```diff
@@ -596,3 +596,18 @@ been enabled:
...
+
+The following command is enabled if an [input_shaper] config section has
+been enabled:
+  - `SET_INPUT_SHAPER [SHAPER_FREQ_X=<shaper_freq_x>]
+    [SHAPER_FREQ_Y=<shaper_freq_y>] [DAMPING_RATIO_X=<damping_ratio_x>]
+    [DAMPING_RATIO_Y=<damping_ratio_y>] [SHAPER_TYPE=<shaper>]
+    [SHAPER_TYPE_X=<shaper_type_x>] [SHAPER_TYPE_Y=<shaper_type_y>]`: Modify
+    input shaper parameters. Note that SHAPER_TYPE parameter resets input shaper
+    for both X and Y axes even if different shaper types have been configured
+    in [input_shaping] section. SHAPER_TYPE cannot be used together with either
```
[the remark this answers is not in the document]
Comment (dmbutyugin) [reply docs/G-Codes.md]: Yes, indeed. I missed this somehow.

**Comment by dmbutyugin** [comment]
> > It is very possible that it's just me, but I don't personally have a good 'hunch' as to which numbers for shaper are 'still good' and which ones are 'too large'. I.e. by just looking at the number, is 0.02 sec too large for EI in terms of smoothing the corners? What about 2hump EI? 3hump EI? I wouldn't be able to tell from the top of my head. I.e. 0.02 sec for EI and 3-hump EI shapers are very not the same. So it may actually mislead the users into thinking that the same period for different shapers == same smoothing, which is wrong.
> 
> I guess I was thinking that it would be easier to document the total "shaping time". So, a ".020s zv shaper" is done over 10ms, ".020s mzv" is 15ms, ".020s ei" is 20ms, ".020s 2hump_ei" is 30ms. Admittedly, it's unclear if ".020s ei" produces more or less rounding than ".040s zv", and it's unclear if there is a consistent "shaping time" that we could broadly say is too high.
> 
> -Kevin

@KevinOConnor 
So, to conclude this discussion: my proposal would be to stick with configuring input shaper parameters in Hz, if you do not have strong objections.

**Comment by KevinOConnor** [comment]
> So, to conclude this discussion: my proposal would be to stick with configuring input shaper parameters in Hz, if you do not have strong objections. I think it is more intuitive to the users: the typical range of 20 - 100 Hz frequency is something that many users are familiar with.

Agreed.

Thanks,
-Kevin

**Comment by KevinOConnor** [comment]
Thanks!

-Kevin

**Comment by kakou-fr** [comment]
is there documentation somewhere that explains how to compute/find the frequency for X Y?

**Comment by Sineos** [comment]
@kakou-fr: Until @dmbutyugin official docu is ready see: https://github.com/dmbutyugin/klipper/blob/scurve-shaping/docs/S-Curve.md#tuning

**Comment by dmbutyugin** [comment]
@Sineos yes, thanks!

@kakou-fr Those docs are in the branch for now, but will be integrated into the mainline in the coming days. For now you can follow the steps in that doc starting from "Tuning" and until and including [Selecting max_accel](https://github.com/dmbutyugin/klipper/blob/scurve-shaping/docs/S-Curve.md#selecting-max_accel)
